File: Java/JDI/jdi-uitest-web/src/main/java/com/epam/jdi/uitests/web/selenium/elements/complex/MultiSelector.java
package com.epam.jdi.uitests.web.selenium.elements.complex;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.commons.EnumUtils;
import com.epam.commons.LinqUtils;
import com.epam.jdi.uitests.core.interfaces.base.IMultiSelector;
import com.epam.jdi.uitests.web.selenium.driver.WebDriverByUtils;
import com.epam.jdi.uitests.web.selenium.elements.apiInteract.GetElementModule;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;

import java.util.List;

import static com.epam.commons.LinqUtils.*;
import static com.epam.commons.PrintUtils.print;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;

/**
 * Created by roman.i on 03.10.2014.
 */

public abstract class MultiSelector<TEnum extends Enum> extends BaseSelector<TEnum> implements IMultiSelector<TEnum> {
    private String separator = ", ";

    public MultiSelector() {
        super();
    }

    public MultiSelector(By optionsNamesLocator) {
        super(optionsNamesLocator);
    }

    public MultiSelector(By optionsNamesLocator, By allLabelsLocator) {
        super(optionsNamesLocator, allLabelsLocator);
    }

    protected void clearAction() {
        if (!hasLocator() && allLabels() == null)
            throw exception("Can't clear options. No optionsNamesLocator and allLabelsLocator found");
        if (getLocator().toString().contains("%s"))
            throw exception("Can't clear options. Specify allLabelsLocator or fix optionsNamesLocator (should not contain '%s')");
        if (allLabels() != null) {
            clearElements(allLabels().getWebElements());
            return;
        }
        List<WebElement> elements = getAvatar().searchAll().getElements();
        WebElement element = elements.get(0);
        if (elements.size() == 1 && element.getTagName().equals("select"))
            if (getSelector().getOptions().size() > 0) {
                getSelector().deselectAll();
                return;
            }
            else throw exception("<select> tag has no <option> tags. Please Clarify element locator (%s)", this);
        if (elements.size() == 1 && element.getTagName().equals("ul"))
            elements = element.findElements(By.tagName("li"));
        clearElements(elements);
    }

    private void clearElements(List<WebElement> els) {
        foreach(where(els, el -> isSelectedAction(el.getText())), WebElement::click);
    }

    protected WebElement getElement(String name) {
        if (!hasLocator() && allLabels() == null)
            throw exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
        if (getLocator().toString().contains("%s"))
            return new GetElementModule(WebDriverByUtils.fillByTemplate(getLocator(), name), this).getElements().get(0);
        if (allLabels() != null)
            return getElement(allLabels().getWebElements(), name);
        return getElement(getElementsFromTag(), name);
    }

    private WebElement getElement(List<WebElement> els, String name) {
        if (els == null)
            throw exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
        List<WebElement> elements = where(els, el -> el.getText().equals(name));
        if (elements != null && elements.size() == 1)
            return elements.get(0);
        throw exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
    }

    protected WebElement getElement(int num) {
        if (!hasLocator() && allLabels() == null)
            throw exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
        if (getLocator().toString().contains("%s"))
            throw exception("Can't get options. Specify allLabelsLocator or fix optionsNamesLocator (should not contain '%s')");
        if (allLabels() != null)
            return getElement(allLabels().getWebElements(), num);
        return getElement(getElementsFromTag(), num);
    }

    private WebElement getElement(List<WebElement> els, int num) {
        if (num <= 0)
            throw exception("Can't get option with num '%s'. Number should be 1 or more", num);
        if (num > els.size())
            throw exception("Can't get option with num '%s'. Found only %s options", num, els.size());
        return els.get(num - 1);
    }

    protected boolean isSelectedAction(String name) {
        return isSelectedAction(getElement(name));
    }

    protected boolean isSelectedAction(int num) {
        return isSelectedAction(getElement(num));
    }

    protected void selectListAction(String... names) {
        foreach(names, this::selectAction);
    }

    protected void selectListAction(int... nums) {
        for (int i : nums) selectAction(i);
    }

    protected String getValueAction() {
        return print(areSelected());
    }

    @Override
    protected void setValueAction(String value) {
        selectListAction(value.split(separator));
    }

    public IMultiSelector<TEnum> setValuesSeparator(String separator) {
        this.separator = separator;
        return this;
    }

    public final void select(String... names) {
        actions.select(this::selectListAction, names);
    }

    public final void select(TEnum... names) {
        select(toStringArray(LinqUtils.select(names, EnumUtils::getEnumValue)));
    }

    public final void select(int... nums) {
        actions.select(this::selectListAction, nums);
    }

    public final void check(String... names) {
        clear();
        select(names);
    }

    public final void check(TEnum... names) {
        clear();
        select(names);
    }

    public final void check(int... nums) {
        clear();
        select(nums);
    }

    public final void uncheck(String... names) {
        checkAll();
        select(names);
    }

    public final void uncheck(TEnum... names) {
        checkAll();
        select(names);
    }

    public final void uncheck(int... nums) {
        checkAll();
        select(nums);
    }

    public final List<String> areSelected() {
        return actions.areSelected(this::getNames, this::isSelectedAction);
    }

    public final void waitSelected(TEnum... names) {
        waitSelected(toStringArray(LinqUtils.select(names, EnumUtils::getEnumValue)));
    }

    public final void waitSelected(String... names) {
        actions.waitSelected(n -> timer().wait(() -> isSelectedAction(n)), names);
    }

    public final List<String> areDeselected() {
        return actions.areDeselected(this::getNames, n -> timer().wait(() -> isSelectedAction(n)));
    }

    public final void waitDeselected(TEnum... names) {
        waitDeselected(toStringArray(LinqUtils.select(names, EnumUtils::getEnumValue)));
    }

    public final void waitDeselected(String... names) {
        actions.waitDeselected(n -> timer().wait(() -> isSelectedAction(n)), names);
    }

    public void clear() {
        invoker.doJAction("Clear Options", this::clearAction);
    }

    public void checkAll() {
        foreach(where(getOptions(), label -> !isSelectedAction(label)), this::selectAction);
    }

}

File: C#.Net/JDI/Web/JDI Web/Selenium/Elements/Complex/MultiSelector.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using JDI_Commons;
using Epam.JDI.Core.Interfaces.Base;
using JDI_Web.Selenium.DriverFactory;
using JDI_Web.Selenium.Elements.APIInteract;
using OpenQA.Selenium;
using static Epam.JDI.Core.Settings.JDISettings;

namespace JDI_Web.Selenium.Elements.Complex
{
    public abstract class MultiSelector<TEnum> : BaseSelector<TEnum>, IMultiSelector<TEnum>
        where TEnum : IConvertible
    {
        private string _separator = ", ";
        
        protected MultiSelector(By optionsNamesLocator = null, By allLabelsLocator = null)
            : base(optionsNamesLocator, allLabelsLocator)
        {
            SelectedNameAction = (m, name) => SelectedElementAction(this, GetWebElement(name));
            SelectedNumAction = (m, num) => SelectedElementAction(this, GetWebElement(num));
            GetValueAction = m => AreSelected().Print();
            GetWebElementFunc = (s, name) =>
            {
                var ms = (MultiSelector<TEnum>) s;
                if (!ms.HasLocator && ms.AllLabels == null)
                    throw Exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
                if (ms.Locator.ToString().Contains("{0}"))
                    return new GetElementModule(ms, ms.Locator.FillByTemplate(name))
                        .WebElements[0];
                if (ms.AllLabels != null)
                    return ms.GetWebElement(AllLabels.WebElements, name);
                return ms.GetWebElement(GetElementsFromTag(), name);
            };
        }

        public override Action<BaseSelector<TEnum>, string> SetValueAction => (c, value) => SelectListNamesAction(this, value.Split(_separator));

        public Action<MultiSelector<TEnum>> ClearAction = m =>
        {
            if (!m.HasLocator && m.AllLabels == null)
                throw Exception("Can't clear options. No optionsNamesLocator and allLabelsLocator found");
            if (m.Locator.ToString().Contains("{0}"))
                throw Exception(
                    "Can't clear options. Specify allLabelsLocator or fix optionsNamesLocator (should not contain '{0}')");
            if (m.AllLabels != null)
            {
                m.ClearElements(m.AllLabels.WebElements);
                return;
            }
            var elements = m.WebAvatar.SearchAll().WebElements;
            if (elements.Count == 1 && elements[0].TagName.Equals("select"))
                if (m.Selector.Options.Any())
                {
                    m.Selector.DeselectAll();
                    return;
                }
                else
                    throw Exception($"<select> tag has no <option> tags. Please Clarify element locator ({m})");
            if (elements.Count == 1 && elements[0].TagName.Equals("ul"))
                elements = elements[0].FindElements(By.TagName("li")).ToList();
            m.ClearElements(elements);
        };

        private void ClearElements(IList<IWebElement> els)
        {
            els.Where(el => SelectedNameAction(this, el.Text)).ForEach(el => el.Click());
        }
        
        private IWebElement GetWebElement(IList<IWebElement> els, string name)
        {
            if (els == null)
                throw Exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
            var elements = els.Where(el => el.Text.Equals(name)).ToList();
            if (elements.Count == 1)
                return elements[0];
            throw Exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
        }

        protected IWebElement GetWebElement(int num)
        {
            if (!HasLocator && AllLabels == null)
                throw Exception("Can't get option. No optionsNamesLocator and allLabelsLocator found");
            if (Locator.ToString().Contains("{0}"))
                throw Exception("Can't get options. Specify allLabelsLocator or fix optionsNamesLocator (should not contain '{0}')");
            if (AllLabels != null)
                return GetWebElement(AllLabels.WebElements, num);
            return GetWebElement(GetElementsFromTag(), num);
        }

        private IWebElement GetWebElement(IList<IWebElement> els, int num)
        {
            if (num <= 0)
                throw Exception($"Can't get option with num '{num}'. Number should be 1 or more");
            if (num > els.Count)
                throw Exception($"Can't get option with num '{num}'. Found only {els.Count} options");
            return els[num - 1];
        }


        protected Action<MultiSelector<TEnum>, IList<string>> SelectListNamesAction = 
            (m, names) => names.ForEach(name => m.SelectNameAction(m, name));
        protected Action<MultiSelector<TEnum>, IList<int>> SelectListIndexesAction =
            (m, nums) => nums.ForEach(num => m.SelectNumAction(m, num));
        
        
        public IMultiSelector<TEnum> SetValuesSeparator(string separator)
        {
            _separator = separator;
            return this;
        }

        public void Select(params string[] names)
        {
            Actions.Select((m, v) => SelectListNamesAction(this, names), names);
        }

        public void Select(params TEnum[] names)
        {
            Select(names.Select(name => name.ToString()).ToArray());
        }

        public void Select(params int[] nums)
        {
            Actions.Select((m, i) => SelectListIndexesAction(this, nums), nums);
        }

        public void Check(params string[] names)
        {
            Clear();
            Select(names);
        }

        public void Check(params TEnum[] names)
        {
            Clear();
            Select(names);
        }

        public void Check(params int[] nums)
        {
            Clear();
            Select(nums);
        }

        public void Uncheck(params string[] names)
        {
            CheckAll();
            Select(names);
        }

        public void Uncheck(params TEnum[] names)
        {
            CheckAll();
            Select(names);
        }

        public void Uncheck(params int[] nums)
        {
            CheckAll();
            Select(nums);
        }

        public IList<string> AreSelected()
        {
            return Actions.AreSelected(m => Names, (m, name) => SelectedNameAction(this, name));
        }

        public void WaitSelected(params TEnum[] names)
        {
            WaitSelected(names.Select(name => name.ToString()).ToArray());
        }

        public void WaitSelected(params string[] names)
        {
            Actions.WaitSelected((m, n) => Timer.Wait(() => SelectedNameAction(this, n)), names);
        }

        public IList<string> AreDeselected()
        {
            return Actions.AreDeselected(m => Names, n => Timer.Wait(() => SelectedNameAction(this, n)));
        }

        public void WaitDeselected(params TEnum[] names)
        {
            WaitDeselected(names.Select(name => name.ToString()).ToArray());
        }

        public void WaitDeselected(params string[] names)
        {
            Actions.WaitDeselected((m, n) => Timer.Wait(() =>SelectedNameAction(this, n)), names);
        }

        public void Clear()
        {
            Invoker.DoJAction("Clear Options", m => ClearAction(this));
        }

        public void CheckAll()
        {
            Options.Where(label => !SelectedNameAction(this, label)).ForEach(label => SelectNameAction(this, label));
        }
        
        public void SelectAll() 
        {
            CheckAll();
        }
        public void UncheckAll()
        {
            Clear();
        }

    }
}
